adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/trivy/TrivyAdapter.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2024 Fraunhofer IEM. All rights reserved.
+ *
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ *
+ * SPDX-License-Identifier: MIT
+ * License-Filename: LICENSE
+ */
+
+package de.fraunhofer.iem.kpiCalculator.adapter.tools.trivy
+
+import de.fraunhofer.iem.kpiCalculator.adapter.AdapterResult
+import de.fraunhofer.iem.kpiCalculator.adapter.KpiAdapter
+import de.fraunhofer.iem.kpiCalculator.adapter.kpis.cve.CveAdapter
+import de.fraunhofer.iem.kpiCalculator.model.adapter.trivy.*
+import de.fraunhofer.iem.kpiCalculator.model.adapter.vulnerability.VulnerabilityDto
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.json.*
+import java.io.InputStream
+import kotlin.math.max
+
+object TrivyAdapter : KpiAdapter<TrivyDto> {
+
+    private val jsonParser = Json {
+        ignoreUnknownKeys = true
+        explicitNulls = false
+    }
+
+    override fun transformDataToKpi(data: Collection<TrivyDto>): Collection<AdapterResult> {
+        return CveAdapter.transformDataToKpi(data.flatMap { it.Vulnerabilities })
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
+    fun dtoFromJson(jsonData: InputStream): TrivyDto {
+        val json =  Json.decodeFromStream<JsonElement>(jsonData)
+
+        if (json is JsonArray)
+            return parseV1(json)
+        else if (json !is JsonObject)
+            throw UnsupportedOperationException("The provided Trivy result is not supported.")
+
+        val schemaVersion = json.get("SchemaVersion")?.jsonPrimitive?.intOrNull
+
+        if (schemaVersion == 2)
+            return parseV2(json)
+
+        throw UnsupportedOperationException("Trivy results for schema version '$schemaVersion' are currently not supported.")
+    }
+
+    private fun parseV1(json: JsonArray) : TrivyDto {
+        val v1dto = jsonParser.decodeFromJsonElement<List<TrivyDtoV1>>(json)
+        val vulnerabilities = createVulnerabilitiesDto(v1dto.flatMap { it.Vulnerabilities })
+        return TrivyDto(vulnerabilities)
+    }
+
+    private fun parseV2(json: JsonObject): TrivyDto {
+        val v2dto = jsonParser.decodeFromJsonElement<TrivyDtoV2>(json)
+        val vulnerabilities = createVulnerabilitiesDto(v2dto.Results.flatMap { it.Vulnerabilities })
+        return TrivyDto(vulnerabilities)
+    }
+
+    private fun createVulnerabilitiesDto(vulnerabilities: Collection<TrivyVulnerabilityDto>) : Collection<VulnerabilityDto>{
+        return vulnerabilities
+            .filter { it.CVSS != null }
+            .map {
```

Comment:
performance hint to achieve the same thing in one loop
```
vulnerabilities.mapNotNull {
    if(it.CVSS == null) {
        null
    }

   ... your code here ...
}
```

map not null allows us to map elements to null, which are then removed from the returned list. I don't think this reduces the readability to much (as Kotlin devs see this pattern regularly) and it reduces the duplicate iteration of the array :)